File: framework/python/test/03/leet.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass
class Input1:
    def __init__(self, value):
        self.value = value

class Input2:
    def __init__(self, text):
        self.text = text

class Output:
    def __init__(self, value, text):
        self.val = value
        self.str = text

class Solution:
    def gen(self, i1: Input1, i2: Input2) -> Output:
        return Output(i1.value, i2.text)

if __name__ == '__main__':
    init_logging()
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().gen)

    # step [3]: setup other options
    work.validator = lambda x, y: x.val == y.val and x.str == y.str
    # work.compareSerial = True
    # work.setArgumentParser(index, (a) => b)
    work.setArgumentParser(0, lambda x: Input1(x))
    work.setArgumentParser(1, lambda x: Input2(x))
    work.resultParser = lambda x: Output(len(x), x[0])
    work.resultSerializer = lambda x: [x.str] * x.val
    # work.resultSerializer = (r) => s
    # work.resultParser = (s) => r
    work.run()

File: framework/java/test/03/Leet.java
import soda.unittest.*;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.*;

import soda.leetcode.*;
import soda.unittest.work.TestWork;

import static soda.unittest.LoggerHelper.logger;

class Input1 {
    public int value;
    public Input1(int value) {
        this.value = value;
    }
}

class Input2 {
    public String text;
    public Input2(String text) {
        this.text = text;
    }
}

class Output {
    public int val;
    public String str;
    Output(int val, String str) {
        this.val = val;
        this.str = str;
    }
}

class Solution {
    public Output gen(Input1 i1, Input2 i2) {
        return new Output(i1.value, i2.text);
    }
}

public class Leet implements Supplier<TestWork> {

    @Override
    public TestWork get() {
        var work = new TestWork(new Solution(), "gen");
        // work.setValidator((e, r) -> {...});
        work.setValidator((Output v1, Output v2) ->
                v1.val == v2.val && (v1.str == v2.str || v1.str != null && v1.str.equals(v2.str)));
        // work.setCompareSerial(true);
        // work.setArgumentParser(index, a -> { ... });
        work.setArgumentParser(0, (Integer v) -> new Input1(v));
        work.setArgumentParser(1, (String v) -> new Input2(v));
        // work.setResultParser(r -> { ... });
        work.setResultParser((List<String> v) -> new Output(v.size(), v.get(0)));
        // work.setResultSerializer(r -> {...});
        work.setResultSerializer((Output v) -> {
            String[] s = new String[v.val];
            for (int i = 0; i < v.val; ++i) {
                s[i] = v.str;
            }
            return s;
        });
        return work;
    }

    public static void main(String[] args) throws Exception {
        new Leet().get().run();
    }

}
